include/revng/Pipeline/ContainerInspector.h

```diff
@@ -0,0 +1,115 @@
+/// specialized with the type of the container in question.
+///
+/// the enumerate and removeTarget methods must be implemented with the
+/// appropriate rules
+///
+///
+template<typename Container>
+class ContainerInspector {
+private:
+  Kind *K;
+
+public:
+  using StaticContainer = llvm::SmallVector<ContainerInspector *, 4>;
+
+  ContainerInspector(Kind &K) : K(&K) {
+    getRegisteredInspectors().push_back(this);
+  }
+  virtual ~ContainerInspector() = default;
+
+  virtual TargetsList
+  enumerate(const Context &Ctx, const Container &ToInspect) const = 0;
+  virtual bool remove(const Context &Ctx,
+                      const TargetsList &Targets,
+                      Container &ToInspect) const = 0;
+
+  static TargetsList
+  enumerateTargets(const Context &Ctx, const Container &ToInspect) {
+    TargetsList ToReturn;
+    for (const ContainerInspector *Inspector : getRegisteredInspectors())
+      ToReturn.merge(Inspector->enumerate(Ctx, ToInspect));
+
+    return ToReturn;
+  }
+
+  static bool removeTarget(const Context &Ctx,
+                           const TargetsList &Targets,
+                           Container &ToInspect) {
+    bool RemovedAll = true;
+    for (const ContainerInspector *Inspector : getRegisteredInspectors())
+      RemovedAll = Inspector->remove(Ctx,
+                                     Targets.filter(*Inspector->K),
+                                     ToInspect)
+                   and RemovedAll;
+
+    return RemovedAll;
+  }
+
+private:
+  static StaticContainer &getRegisteredInspectors() {
+    static StaticContainer List;
+    return List;
+  }
+};
+
+/// A inspectable container is a container that delegates the enumeration and
+/// remova of its contents to inspectors, which can be dinamically registered
+/// at runtime.
+///
+/// This facility is needed to be able to decouple the Kind declaration from
+/// the containers declaration, since if it was always containers to name which
+/// kinds can be contained in them, then they would not be able to name kinds
+/// declared downstream libraries. Similarly, should be always kind to be able
+/// to enuemrate the context of a contaienr, then containers declared in
+/// downstream libraries would not be able to use kinds declared upstream.
+///
+/// You should declare a new inspectable container if and only if you think that
+/// downstream user may wish to extend how the enumeration of this container
+/// happens.
+template<typename Derived>
+class InspectableContainer : public Container<Derived> {
+protected:
+  Context *Ctx;
+
+public:
+  InspectableContainer(Context &Ctx) : Container<Derived>(), Ctx(&Ctx) {}
+  ~InspectableContainer() override = default;
+
+  TargetsList enumerate() const final {
+    return ContainerInspector<Derived>::enumerateTargets(*Ctx, *self());
+  }
+
+  bool remove(const TargetsList &Targets) final {
+    return ContainerInspector<Derived>::removeTarget(*Ctx, Targets, *self());
+  }
```

Comment:
well, one has to look up inside the static list. we can push this inside the container but there is not much difference. This is never used anywhere else anyway.

Reply:
This confuses me.
There is a one-to-many relationship between `InspectableContainer<X>` and classes derived from `ContainerInspector<X>`. However `ContainerInspector<X>` contains a `static` list of all the inspectors. Wouldn't be more logical to move all the static fields and methods in `InspectableContainer<X>`?

This would probably require you to push `ContainerInspector::ContainerInspector` outline, but it seems more logical to me.